File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    bool dfs(TreeNode* curr, int left, int right) {
        if (!curr) { return true; }

        if ( !(left < curr->val && curr->val < right) ) {
            return false;
        }

        return dfs(curr->left, left, curr->val) && dfs(curr->right, curr->val, right);
    }
public:

    bool isValidBST(TreeNode* root) {
        return dfs(root, INT_MIN, INT_MAX); //bounds
    }
};


/* 
-infinity < root_val < infinity
-infinity < 3 < 5
5 < 7 < infinity
5 < 4 < 7 ERRR kaboom
7 < 8 < infinity

            5
           / \
          3   7
             /\
            4  8
*/

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def dfs(self, root: Optional[TreeNode]):
        if not root:
            return 0

        left = self.dfs(root.left)
        right = self.dfs(root.right)

        self.best = max(self.best, left + right)
        
        return 1 + max(left, right)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.best = 0 # "global"
        self.dfs(root)
        return self.best


"""
dfs

at eeach node caulacte left + right height

"""